### ISSUE TITLE
BeEqualTo should support deeper compare

### ISSUE THREAD

**Issue by noam-almog** [etorreborre/specs2#341]
Hi,

Specs2 output in case of equality is lacking if you are matching two objects.

You will get a full output that says 

```
Expected: ClassA(val1.. valN) 
Actual: ClassA(val1'... valN')
```

without you actually knowing what is the difference. In case of random values and more than two properties it gets really hard to actually know what the difference.

I want to PR specs2 with a change that will give you a more meaningful response with the actual properties that are different.

For example:

```
Expected: ClassA(val1,...valN)
Actual: ClassA(val1,...valN)
Differences:
property [val1] Expected: someVal, Actual: anotherVal
property[valX] Expected: someVal, Actual: anotherVal
```

I'm open to suggestions about the format.
- Noam

**Comment by etorreborre** [comment]
I think that we should restrict this to case classes, so probably to classes extending `Product` but we need to make sure that this doesn't clash with things like `Tuples`.

Then if we do a deep comparison we need to make sure that we display failures only for the leaves and not at each level. Other than this I guess that we can reuse most of the existing machinery for comparisons (on lists if your attribute is a `List` etc...).

Probably we can add a new `Details` subclass (see `Result.scala`): `ProductDetails` to hold the differences and extend the `Diffs` trait to handle this case (and do a default implementation in `SmartDiffs`).

**Comment by andriuss** [comment]
:thumbsup:

**Comment by mdedetrich** [comment]
I wouldn't recommend adjusting `beEqualTo` to use anything other than the standard `==` check (and if you want to do deep equality checking, should probably make a `beDeepEqualTo` or something like that).

Many people still use `==` to test for equality, and `beEqualTo` is a very handy method to test if an `override equals(a: Any) actually works as expected.

**Comment by etorreborre** [comment]
I should probably provide a new set of matchers for `scalaz` and for `cats` using their respective `Eq` typeclasses.

**Comment by etorreborre** [comment]
This issue would be best solved by using a matcher delegating the diff to https://github.com/xdotai/diff. I have sketched such a matcher [there](https://gist.github.com/etorreborre/5699231208c47ee78d6eee0f015e5a1b) (but got some error I don't yet understand).

**Comment by noam-almog** [comment]
@etorreborre 
I'm starting to work on a PR for this issue and I wanted to get some feedback. 
I will try to incorporate some of the ideas from xdotai's code.

The basic outline of this would be:

Primitives:
left -> right, for example: 1 -> 2

Traversable:
Seq(a, b, c -> d)
In case of long lists we can reduce the printing to ... to improve readability
Seq(..., b, c -> d, e, ....)

Objects:
use named parameters:
ClassA(propA: "a", propB: "b" -> "c")

**Comment by etorreborre** [comment]
That looks good to so far except maybe for one point:
```scala
case class Box[A](a: A)

sealed trait Thing
case class Fruit(name: string) extends Thing
case class Paper(area: Double) extends Thing
```

How do you represent the diff between:
```scala
Box(Fruit("apple")): Box[Thing]
Box(Paper(20.0)): Box[Thing]
```

If you display `Fruit("apple") -> Paper(20.0)` this could be a very large thing to display.

**Comment by noam-almog** [comment]
@etorreborre 
it's actually supposed to be longer :)
```
Not equal: Box(a: Fruit("apple") -> Paper(20.0))
```
But there is no real way around this, with this format at least you know the name of the property which is different.
Do you have any suggestion on how to approach this ? perhaps better indentation of the message ?

**Comment by etorreborre** [comment]
Not really. Maybe just an option to do short or expanded diffs:
```scala
// short
Not equal: Box(a: Fruit(...) -> Paper(...))

// expanded
Not equal: Box(a: Fruit("apple") -> Paper(20.0))
```

**Comment by noam-almog** [comment]
In the case that you wrote it's enough to show that the types are different, so toString in this case would be redundant. same goes for long lists, showing all the members is redundant as well, we can save some space there as well.
but if comparing a deep tree of objects you can get this:
```
   A(a = B(b = C(c= D(d = "x" -> "y", ...) ...) ...), zz = 4 -> 5 )
```
I want to add another thing, what do you do with lists that has too many or too few elements ?
Right now we want to do something like this:
```
   Seq(1, 2, 3, missing: 4, 5, 6)
   Seq(1, 2, 3, added: 4, 5, 6)
```
also here, we can shorten the list with:
```
   Seq(..., 4, 5 -> 666, 6, ...)
```
btw, I think if I can make the abbreviation readable, if someone wants to get the full result he can always look at the Expected, Actual clause.

**Comment by etorreborre** [comment]
I agree, `added/missing` is really useful.

**Comment by etorreborre** [comment]
BTW see [here](https://github.com/ambiata/disorder.hs/pull/94) for another example of display. This is in Haskell but is probably translatable to Scala with case classes.

**Comment by noam-almog** [comment]
I'll try to commit this in a few stages so that you can play with it.
pretty print and abbreviation will come at a later commits to the PR.

bwt, the original code used colors, I also noticed that specs support coloring in some implementations. I can also add coloring, but I couldn't figure out how to add those hints to the matcher. my question is, can I suggest colors from the matcher code ? and if so, how ?

**Comment by etorreborre** [comment]
I finally merged your code with some modifications into `master` (yay!). I extended the `ComparisonResult` class with a `def identical: Boolean` method in order to be able to reuse the `Diffable` typeclass as an "Equality" typeclass, and let users different other ways of comparing their objects.

I also added some indenting on case class comparison for better readability. We will need to see if that scales to real-life examples.

For now this is all available in `3.8.7-20170130231531-f808b61`. I will probably publish as `3.8.8` soon.

Thanks again for your work and being so patient, almost 2 years since the issue was first created :-)
